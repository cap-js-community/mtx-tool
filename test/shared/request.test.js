"use strict";
const { request } = require("../../src/shared/request");

const mockFetchLib = require("node-fetch");
jest.mock("node-fetch", () => jest.fn());

const { Logger: MockLogger } = require("../../src/shared/logger");
const { outputFromLogger } = require("../test-util/static");
const mockLogger = MockLogger.getInstance();
jest.mock("../../src/shared/logger", () => require("../__mocks/shared/logger"));

const baseOkResponse = {
  ok: true,
  status: 200,
  statusText: "OK",
};

const baseBadRequestResponse = {
  ok: false,
  status: 400,
  statusText: "Bad Request",
  text() {
    return "failure reason";
  },
};

describe("request", () => {
  test("basic ok", async () => {
    mockFetchLib.mockReturnValueOnce(baseOkResponse);
    await request({ url: "https://server", pathname: "/path" });
    expect(mockFetchLib.mock.calls).toMatchSnapshot();
    expect(outputFromLogger(mockLogger.info.mock.calls)).toMatchInlineSnapshot(
      `"GET https://server/path 200 OK (0ms)"`
    );
  });

  test("basic bad request", async () => {
    mockFetchLib.mockReturnValueOnce(baseBadRequestResponse);
    await expect(request({ url: "https://server", pathname: "/path" })).rejects.toMatchInlineSnapshot(`
            [Error: got bad response 400 from https://server/path
            failure reason]
          `);
    expect(mockFetchLib.mock.calls).toMatchSnapshot();
    expect(outputFromLogger(mockLogger.info.mock.calls)).toMatchInlineSnapshot(
      `"GET https://server/path 400 Bad Request (0ms)"`
    );
  });

  test("bad request unchecked", async () => {
    mockFetchLib.mockReturnValueOnce(baseBadRequestResponse);
    await request({ url: "https://server", pathname: "/path", checkStatus: false });
    expect(mockFetchLib.mock.calls).toMatchSnapshot();
    expect(outputFromLogger(mockLogger.info.mock.calls)).toMatchInlineSnapshot(
      `"GET https://server/path 400 Bad Request (0ms)"`
    );
  });

  test("ok with no logging", async () => {
    mockFetchLib.mockReturnValueOnce(baseOkResponse);
    await request({ url: "https://server", pathname: "/path", logged: false });
    expect(mockFetchLib.mock.calls).toMatchSnapshot();
    expect(mockLogger.info).toHaveBeenCalledTimes(0);
  });

  test("ok with basic auth", async () => {
    mockFetchLib.mockReturnValueOnce(baseOkResponse);
    await request({ url: "https://server", pathname: "/path", auth: { username: "username", password: "password" } });
    expect(mockFetchLib.mock.calls).toMatchSnapshot();
  });

  test("ok with bearer auth", async () => {
    mockFetchLib.mockReturnValueOnce(baseOkResponse);
    await request({ url: "https://server", pathname: "/path", auth: { token: "token" } });
    expect(mockFetchLib.mock.calls).toMatchSnapshot();
  });
});
