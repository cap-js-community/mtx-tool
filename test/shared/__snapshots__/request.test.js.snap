// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`request tests bad request unchecked 1`] = `
[
  [
    "https://fake-server.com/path",
    {
      "headers": {
        "X-Correlation-Id": "88888888-8888-8888-8888-888888888888",
        "X-CorrelationId": "88888888-8888-8888-8888-888888888888",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`request tests basic bad request 2`] = `
[
  [
    "https://fake-server.com/path",
    {
      "headers": {
        "X-Correlation-Id": "88888888-8888-8888-8888-888888888888",
        "X-CorrelationId": "88888888-8888-8888-8888-888888888888",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`request tests basic ok 1`] = `
[
  [
    "https://fake-server.com/path",
    {
      "headers": {
        "X-Correlation-Id": "88888888-8888-8888-8888-888888888888",
        "X-CorrelationId": "88888888-8888-8888-8888-888888888888",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`request tests ok with basic auth 1`] = `
[
  [
    "https://fake-server.com/path",
    {
      "headers": {
        "Authorization": "Basic dXNlcm5hbWU6dGVzdC1wYXNzd29yZA==",
        "X-Correlation-Id": "88888888-8888-8888-8888-888888888888",
        "X-CorrelationId": "88888888-8888-8888-8888-888888888888",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`request tests ok with bearer auth 1`] = `
[
  [
    "https://fake-server.com/path",
    {
      "headers": {
        "Authorization": "Bearer test-token",
        "X-Correlation-Id": "88888888-8888-8888-8888-888888888888",
        "X-CorrelationId": "88888888-8888-8888-8888-888888888888",
      },
      "method": "GET",
    },
  ],
]
`;

exports[`request tests ok with lots of options 1`] = `
[
  [
    "https://server/path?hello=world&foo=bar#hashed",
    {
      "headers": {
        "X-Correlation-Id": "88888888-8888-8888-8888-888888888888",
        "X-CorrelationId": "88888888-8888-8888-8888-888888888888",
      },
      "method": "GET",
      "redirect": true,
    },
  ],
]
`;

exports[`request tests ok with no logging 1`] = `
[
  [
    "https://fake-server.com/path",
    {
      "headers": {
        "X-Correlation-Id": "88888888-8888-8888-8888-888888888888",
        "X-CorrelationId": "88888888-8888-8888-8888-888888888888",
      },
      "method": "GET",
    },
  ],
]
`;
